### DISCUSSION TITLE
extremely high number of input tokens

### DISCUSSION THREAD

**Discussion by chrisadas** [open-webui/open-webui#7281] [Issues]
I have been using open-webui for a while, but more heavily lately. I notice that the LLM APIs I use (OpenAI, Anthropic, Mistral, all through LiteLLM proxy) are reporting very high number of input tokens and I cannot reconcile this with my usage.

Example: In 1 session of using open-webui for creating writing, I have used 1.5M input tokens and 13k output tokens, as reported by Mistral dashboard.

![image](https://github.com/user-attachments/assets/c57f0b58-8898-45ae-9633-1b15d499050e)

My usage have the LLM continuing to write an existing story, 8-12 paragraphs at a time. There is a long existing story present in the chat. But with the default context_length = 2048 tokens, it should not be sending the whole chat history as input context, right?

I am using knowledge document, but they are 3.6kb in size so I also don't think it explains the number of input tokens.

Does anyone has idea what is happening? Or how I can diagnose? Is there a way to inspect the whole LLM calls made by open-webui?

**Comment by ImLuke954** [comment]
I have this exact same issue while using OpenAI API. Have you found any solution to this?

**Comment by adamoutler** [comment]
This could be the problem where OpenWebUI gets lagged during pipes. Just a simple word "Test" is detected as 4250 input tokens. 

![Screenshot_20250104-085335](https://github.com/user-attachments/assets/ba10b8fc-1a3c-4757-a1d1-de1d8c728b95)

**Reply by roryeckel** [reply]
That is per second, the request may have finished in milliseconds

**Comment by ramezquitao** [comment]
I'm having the same issue with both OpenAI and Anthropic. Could this be a configuration issue?

**Comment by tjbck** [comment]
`context_length` is a param for Ollama and has nothing to do with the chat history.

**Comment by chrisadas** [comment]
@tjbck I think this has caused confusion for myself and a few others because the tool tip of `context_length` just say `Sets the size of the context window used to generate the next token. (Default: 2048)` with no mention of Ollama, and this parameter is visible for user to change for all models not just Ollama ones.

@ImLuke954 @ramezquitao I add this function and enable it in the chat/model that I use. This one truncate the context sent to API by number of chat message instead of number of tokens. (There are other function available that does this by number of tokens as well.) The `n_last_messages` parameter is configurable. For some API this needs to be an odd number to work.

https://openwebui.com/f/hub/context_clip_filter

**Reply by ImLuke954** [reply]
Thanks mate. This function works.

context_length has no effect on API models apparently. Even after setting it, it sends infinite input tokens. In a single, long chat I wasted 1.3M input token because context_length didn't do what it was expected to do.

**Comment by ramezquitao** [comment]
@chrisadas, thanks for the function. One last question: By default, in openwebui how many tokens or chat lines are sent to the model as context?

**Comment by prasoon2211** [comment]
Having the same problem. Any fixes?

**Comment by GindaChen** [comment]
Also encountered this. I think this will be potentially one of the biggest problem stopping people from using open-webui.

**Comment by adamoutler** [comment]
Could it be memories?

**Comment by rgaricano** [comment]
Maybe this project help for conta: https://github.com/VariantConst/OpenWebUI-Monitor
(I did't try it, I only use local resources)

**Comment by xiaoyanggu** [comment]
I encounter this problem too. What I found is that open webui has auto gen for Title, Tag, and user message autocomplete features turned on by default.  Also by default they use 'Current Model' for those purposes, which causes substantial (orders of magnitude higher) input token usage on the model. You would want to change that to avoid sending excessive token to your expensive models.

<img width="721" alt="image" src="https://github.com/user-attachments/assets/4e6e0bc9-8ea5-4438-adc7-c549ebef9edb" />

**Reply by xiaoyanggu** [reply]
<img width="762" alt="image" src="https://github.com/user-attachments/assets/a3a2a2c8-dd35-4133-b0dd-22ae9624a93d" />

**Comment by feluxe** [comment]
I think it happened to me with gpt-4o-latest as well. I just burned ~4M tokens.

![2025-03-06T00-58-15_screenshot](https://github.com/user-attachments/assets/7fac9323-628d-4eea-b1b7-fec903590209)

![2025-03-06T00-53-51_screenshot](https://github.com/user-attachments/assets/082e7222-9b0c-448d-aeaa-0cbbca1c0e0e)

I used the default settings for pretty much everything when it happened.

It was a coding session, where I copied a bash script (~300 LOC) and asked GPT to add some features and help debug it. The chat got quite long, but my questions were rather short.

**Reply by spammenotinoz** [reply]
Well that is how API's work, each new message includes the entire history for context.
The filter posted above is good, as it preserves the system message and first user message.

I personally like the Open-WebUI approach.  I have seen other cost focus approaches where the history is sent to a low cost model first to summarise before sending to the chosen model, great to reduce cost, but personally am not a fan.
The one that I do like, is the Prompt Caching.  I have not been able to get Prompt Caching working with Open-Webui.

**Comment by Seviks** [comment]
I had the same problem, after I tampered with context length for OpenAi models. Since I set it back to default (2048 I think) its gone.
If this is truly the cause, I think I have to define separate models so we can choose depending on the needs.

**Comment by magranger16** [comment]
I'm having this problem as well. gpt-4o only allows 30,000 input tokens. Changing the context size doesn't do anything for me, turning off the memory only reduces the input tokens sent by a small amount, and starting a new chat with the word "test" only shows that 4 tokens were sent when I ask the ai.

I can't figure out what's causing the issue, I've restarted the docker container, checked the audit.log file, tried looking at the docker debug log, but I can't figure it out. I also tried out the Function mentioned above and it didn't change anything.

**Reply by spammenotinoz** [reply]
> I'm having this problem as well. gpt-4o only allows 30,000 input tokens. Changing the context size doesn't do anything for me, turning off the memory only reduces the input tokens sent by a small amount, and starting a new chat with the word "test" only shows that 4 tokens were sent when I ask the ai.
> 
> I can't figure out what's causing the issue, I've restarted the docker container, checked the audit.log file, tried looking at the docker debug log, but I can't figure it out. I also tried out the Function mentioned above and it didn't change anything.

Sorry but you are not clear on what your issue is.
Gpt-4o supports a combined limit of 128k with an output cap of 4k, so the input cap is 124k.  (If you are using another provider not OpenAI then they may impose other caps).
You may be confusing the rate limit with token limit, depending on your tier a rate limit of 30k/second may apply to your account.
Either way the response back from OpenAI always makes it clear.

If you concern is the high cost, that is how API's work, each NEW message in a conversation sends back the entire conversation history.  That's called an exponential increase in API usage as the conversation progresses.  As posted above there are functions to address this, just make sure they are async.
That's about the only downside of this project the lack of support for Caching, so your costs are 90% more than alternate projects.  But given the low price of API's I don't think it's a major issue for many.
If you do need Prompt Caching, I believe there are Pipelines that support it, but have not tried.

Sorry should also mention, that Prompt Caching isn't just about cost, the main payback is token management.
Typically you can input 50-65% more tokens for much more productive outcomes and less likely to hit token limits.

**Comment by alanalston** [comment]
I just burned through 16m tokens uploading relatively small files. Happens on OpenAI and Gemini; across the board then. 

> The input token count (1321595) exceeds the maximum number of tokens allowed (1048576).


![image](https://github.com/user-attachments/assets/635ab82d-9a57-4896-b995-e46351e37459)

**Comment by spammenotinoz** [comment]
Like me you probably have by-pass embeddings
enabled.  Really not an issue if you use low cost models.  You need some controls to stop blow-outs with expensive models.

**Comment by spammenotinoz** [comment]
Ohh.. should remind you the way API’s work, as they have no memory each call sends back the entire conversation history so token count increases significantly as the conversation progresses.  Write your own filter, to counter this.  Many different strategies and that’s what’s great out Open-Webui you can tailor to you needs.

**Comment by Vsono** [comment]
I had a similar problem. Simple inputs like 'hi' were consuming 4k~6k tokens.
Disable the builtin tools in each model's settings, or enable only the ones you need. (even better, make a clone with the tools disabled).

Here are the rough token counts added **just** by enabling each builtin tool for my case:
| Tool | Tokens |
| -------- | -------- | 
| Time & Calculation  | 500  | 
| Memory | 0 |  
| Chat History | 500 |  
| Notes | 700 |  
| Knowledge Base | 1700 |  
| Channels | 0 |  
| Web Search | 0 |  
| Image Generation | 0 |  
| Code Interpreter | 0 |  
| Task Management | 500 |  
| Automations | 1100 |  
| Calendar | 1400 |  

Your numbers may vary. For example, if I enable Web Search in the chat UI, that adds 400 tokens instead of 0. Note that I also have Memory disabled in the global settings, so that could add another in your case.

I think they should put some warnings because for some models those can get very expensive.

**Comment by Vsono** [comment]
Oh, also disable Title Auto-Generation, Follow-Up Auto-Generation, Chat Tags Auto-Generation (in Settings - Interface). Each of those can add a lot of tokens, like tens of thousands tokens after EACH message.

Or, change the task model to a cheaper one in Admin Panel - Settings - Interface

**Comment by marcoshimselfxd** [comment]
> I have been using open-webui for a while, but more heavily lately. I notice that the LLM APIs I use (OpenAI, Anthropic, Mistral, all through LiteLLM proxy) are reporting very high number of input tokens and I cannot reconcile this with my usage.
> 
> Example: In 1 session of using open-webui for creating writing, I have used 1.5M input tokens and 13k output tokens, as reported by Mistral dashboard.
> 
> ![image](https://private-user-images.githubusercontent.com/16176343/389199788-c57f0b58-8898-45ae-9633-1b15d499050e.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODczNjA3NTQsIm5iZiI6MTc4NzM2MDQ1NCwicGF0aCI6Ii8xNjE3NjM0My8zODkxOTk3ODgtYzU3ZjBiNTgtODg5OC00NWFlLTk2MzMtMWIxNWQ0OTkwNTBlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA4MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwODIyVDAxMDA1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ0NGRiZDA4YTEzZjI2YTFkMzQ1OTRiZTdhZTE0MzA5YjI0NjBlMmUyMmFmNTAzYzdkMDkzNDFmYjY3NjFjMTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.9HO5h5dJsO-Zi4dth2dB740zVXvpOhZYbAa_hgne19w)
> 
> My usage have the LLM continuing to write an existing story, 8-12 paragraphs at a time. There is a long existing story present in the chat. But with the default context_length = 2048 tokens, it should not be sending the whole chat history as input context, right?
> 
> I am using knowledge document, but they are 3.6kb in size so I also don't think it explains the number of input tokens.
> 
> Does anyone has idea what is happening? Or how I can diagnose? Is there a way to inspect the whole LLM calls made by open-webui?

Have you found a fix for this?

Noticed the same behavior here, I noticed DS burning loooooots of tokens

**Comment by umutdenizNB** [comment]
Worth separating the causes here — history growth, bypassed embeddings and the auto-title/tags task calls are all separate from the builtin-tool overhead in the table above. This only addresses the tool slice.

For tools you don't need, disabling per model is the right fix. For ones you need occasionally — knowledge base at ~1700 tokens is usually the one people can't just switch off — there's a middle option: keep them enabled but hide the schemas until the model asks.

Filter: https://github.com/anchor-cc/openwebui-dynamic-mcp — the model sees one `load_tools` tool plus the hidden tool names. On my setup that took ~29k prompt tokens per call down to ~600 on the first call, at the cost of one extra round trip the first time a chat uses a tool.
